Support Kafka 0.10.1.X and 0.10.2.1

JensRantil opened the pull request:
Fixes #24.

JensRantil committed:
fea(kafka): support for 0.10.2.1 format

JensRantil committed:
log(kafka): improve error somewhat

Helps debugging the breaking test, but will also be nice when debugging
this in the future.

stijndehaes commented:
@JensRantil I would really like this to get merged can I help test this or something? :)

JensRantil commented:
@stijndehaes I'll have a look at this over the weekend. Most likely, we will have to have a CLI flag that configures which format to parse.

cl0udgeek commented:
this would be awesome! Also, 0.11 got release too so hopefully this will work with that one too

JensRantil committed:
ref(interfaces): array elements must not be nil

...and since we don't really have any performance reasons to have the
nilable, I'm simplifying a bunch of types to avoid pointer references.

This was triggered by me about to introduce multiple different parsers.

JensRantil committed:
ref(collector): split into separate parsing file

This is a prerequisite to be able to support multiple parsers.

JensRantil committed:
ref(kafka): introduce interface for parser

Will introduce different parsers.

JensRantil committed:
ref(kafka): introduce `DelegatingParser`

...and back it by two different Kafka versions. This is a preparatory
commit to introduce support for the Kafka 0.10.2.1 format.

JensRantil committed:
fix(import): fix case-sensitive import

JensRantil commented:
@stijndehaes @k1ng87 This is now ready for testing. I managed to not have to require a new parameter by using the header as a fingerprint for which version is running. Any of you guys who'd be willing to build from this branch and give it a spin? Unfortunately I don't have a 0.10.2.1 cluster available to test on.

cl0udgeek commented:
yeah...I take a look at it tomorrow....thanks for the change!

JewelPengin commented:
@k1ng87 and I have been testing this. Unfortunately we were testing against 0.11.0 which, while it has the same format as 0.10.2.1, it introduces some extra lines at the beginning of the admin command (see https://github.com/apache/kafka/blob/0.11.0/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L62-L65). And in all actuality, 0.10.2.1 has the same (https://github.com/apache/kafka/blob/0.10.2/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L61-L64). I'll try to get a 0.10.2 or a 0.10.1 instance to test on.

Here's a full example of what I see:
```
>$ kafka-consumer-groups --bootstrap-server localhost:9092 --list
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).

test-consumer-group
>$ kafka-consumer-groups --bootstrap-server localhost:9092 --describe --group test-consumer-group
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).


TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
test                           0          19              19              0          consumer-1-c00428b4-8a62-4a1f-a9f7-8d4b71640ef1   /10.100.100.10                 consumer-1
```

JensRantil committed:
ref(kafka): use static variables

...instead of constructing the same immutable objects every time.

JensRantil committed:
ref(kafka): better function name

More clear what it does.

JensRantil commented:
@JewelPengin Sorry for such a late feedback here - and thanks a lot for testing this out! Would it be possible for you to see if the first two lines are stderr or stdout? Currently we combine the two when reading output, but maybe simply focusing on stdout will do.

JewelPengin commented:
No worries! The lines are printed to stderr. Only thing is that we'd be missing out on two other errors in this process. The `Consumer group '$groupId' has no active members.` and `Warning: Consumer group '$groupId' is rebalancing.` Everything else goes to stdout. Kinda inconsistent on their part 😒

JensRantil committed:
fix(kafka 0.10.2.1): split stderr/stdout

Got input from JewelPengin that Kafka CLI command exposes some stderr
every time.

JensRantil committed:
ref(kafka): better test names

More grammatically correct.

JensRantil committed:
tst(kafka): test parsing stderr for Kafka 0.10.2.1

JensRantil commented:
@JewelPengin Would you mind giving this a spin? Does tests in https://github.com/kawamuray/prometheus-kafka-consumer-group-exporter/pull/29/commits/cef2a19504e5842a77015ee611db5ec90b81b1f6 look sane?

vavrusa commented on kafka/parsing.go:
```diff
@@ -0,0 +1,183 @@
+	var err error
+
+	var currentOffset int64
+	currentOffset, err = parseLong(matches[p.indexByName["currentOffset"]])
```
This will crash with the 0.10.1 format:

```
panic: runtime error: index out of range

goroutine 40 [running]:
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*regexpParser).parseLine(0xc4200b0f20, 0xc420186000, 0x0, 0xc4201b47d0, 0x0, 0x0)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/parsing.go:89 +0x654
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*regexpParser).Parse(0xc4200b0f20, 0xc420186000, 0x3177, 0x0, 0x0, 0x0, 0x0, 0x0, 0xad2180, 0xc42012a860)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/parsing.go:70 +0x271
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*DelegatingParser).Parse(0xc420112ac0, 0xc420186000, 0x3177, 0x0, 0x0, 0x3, 0xc420186000, 0x3177, 0x0, 0x0)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/parsing.go:172 +0x8f
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka.(*ConsumerGroupsCommandClient).DescribeGroup(0xc420100960, 0xad8d80, 0xc42013ccc0, 0xc42012c8a0, 0x12, 0x0, 0x0, 0x0, 0x0, 0x0)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/kafka/collector.go:79 +0x149
github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync.(*FanInConsumerGroupInfoClient).describeLoop.func1(0xc4200116c0, 0xc42012c8a0, 0x12, 0xc42013cd80, 0xad8d80, 0xc42013ccc0, 0xc42013ce40)
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync/metrics.go:159 +0x71
created by github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync.(*FanInConsumerGroupInfoClient).describeLoop
	/src/github.com/kawamuray/prometheus-kafka-consumer-group-exporter/sync/metrics.go:164 +0x178
```

JensRantil replied:
Thanks! Would it be possible for you to send me a sample of the output from the `kafka-consumer-groups.sh` script of `0.10.1.X` release? If you could separate `stderr` from `stdout` that would be nice.

vavrusa replied:
Sure:

```sh
# ./kafka-consumer-groups.sh --describe --group group1 --bootstrap-server bootstrap:9092
GROUP                          TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             OWNER
group1             requests                       11         1176417636761   1176419294709   1657948         consumer1_/1.1.1.5
group1             requests                       25         1144248140115   1144249685348   1545233         consumer1_/1.1.1.18
group1             requests                       24         1101749049926   1101751118906   2068980         consumer1_/1.1.1.23
group1             requests                       16         1065240411131   1065242284398   1873267         consumer1_/1.1.1.6
group1             requests                       81         76858987875     76860609239     1621364         consumer1_/1.1.1.18
group1             requests                       68         1182773827958   1182897904075   124076117       consumer1_/1.1.1.19
group1             requests                       66         1058973733081   1058974821719   1088638         consumer1_/1.1.1.12
group1             requests                       30         1156698657130   1156700015952   1358822         consumer1_/1.1.1.10
...
```

JensRantil replied:
@vavrusa Thanks a lot for that test case. I believe we now also have support for 0.10.1.X 🎉

JensRantil committed:
ref(parsing): use subtests

...to more easily know what failed.

JensRantil committed:
tst(collector): correct assertion error

Had a few typos.

JensRantil committed:
fea(parsing): add support for 0.10.1.X

Cudos to @vavrusa for giving me a testcase in [1].

[1] https://github.com/kawamuray/prometheus-kafka-consumer-group-exporter/pull/29#discussion_r136854490

JensRantil commented:
My current gig recently upgraded Kafka to 0.10.2.1. I'll give this PR a spin and will merge if things are looking promising.

JensRantil committed:
tst(parsing): more realisting input

Includes proper consumer and address.

JensRantil committed:
fix(parsing): avoid panicing if fields are missing

Instead, log it.

JensRantil committed:
fix(parsing): ignore empty lines

JensRantil committed:
doc(readme): update supported versions

0xdavidm commented:
@JensRantil Do you think that you would merge this today so we can test it?

JensRantil commented:
@panda87 Done ;)

JensRantil commented:
...also, please file new issues if you'd discover anything not working as expected.

0xdavidm commented:
Wow, that's was quick!
